src/stacked_borrows/diagnostics.rs

```diff
@@ -0,0 +1,338 @@
+    },
+    Untagged {
+        recently_created: Option<(AllocRange, SpanData)>,
+        recently_invalidated: Option<(AllocRange, SpanData)>,
+        matching_created: Option<(AllocRange, SpanData)>,
+        protected: Option<(SbTag, SpanData, SpanData)>,
+    },
+}
+
+pub trait GlobalStateExt {
+    fn current_span(&self, threads: &ThreadManager<'_, '_>) -> Span;
+
+    fn log_creation(
+        &mut self,
+        parent: Option<SbTag>,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn log_invalidation(
+        &mut self,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn log_protector(
+        &mut self,
+        orig_tag: SbTag,
+        tag: SbTag,
+        alloc: AllocId,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn get_stack_history(
+        &self,
+        tag: SbTag,
+        alloc: AllocId,
+        alloc_range: AllocRange,
+        offset: Size,
+        protector_tag: Option<SbTag>,
+    ) -> Option<TagHistory>;
+}
+
+impl GlobalStateExt for GlobalStateInner {
+    fn current_span(&self, threads: &ThreadManager<'_, '_>) -> Span {
+        threads
+            .active_thread_stack()
+            .into_iter()
+            .rev()
+            .find(|frame| {
+                let def_id = frame.instance.def_id();
+                def_id.is_local() || self.local_crates.contains(&def_id.krate)
+            })
+            .map(|frame| frame.current_span())
+            .unwrap_or(rustc_span::DUMMY_SP)
+    }
+
+    fn log_creation(
+        &mut self,
+        parent: Option<SbTag>,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
+    ) {
+        let span = self.current_span(threads);
+        let extras = self.extras.entry(alloc).or_default();
+        extras.creations.push(Event { parent, tag, range, span, time: extras.current_time });
+        extras.current_time += 1;
+    }
+
+    fn log_invalidation(
+        &mut self,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
+    ) {
+        let span = self.current_span(threads);
+        let extras = self.extras.entry(alloc).or_default();
+        extras.invalidations.push(Event {
+            parent: None,
+            tag,
+            range,
+            span,
+            time: extras.current_time,
+        });
+        extras.current_time += 1;
+    }
+
+    fn log_protector(
+        &mut self,
+        orig_tag: SbTag,
+        tag: SbTag,
+        alloc: AllocId,
+        threads: &ThreadManager<'_, '_>,
+    ) {
+        let span = self.current_span(threads);
+        let extras = self.extras.entry(alloc).or_default();
+        extras.protectors.push(Protection { orig_tag, tag, span });
+        extras.current_time += 1;
+    }
+
+    fn get_stack_history(
```

Comment:
the use of iterator combinators vs for loops and `and_then`/`map` vs `if let` feels very random in this function and I keep expecting a specific choice to have a reason.

Maybe create one function per `TagHistory` field and use `?` more within these functions?